README.md
# Collage

Collage is a simple-to-use image processing library for Clojure. It's intended
to be a drop-in solution for high-level image processing needs. It draws some
inspiration from Python's [PIL](http://effbot.org/imagingbook/pil-index.htm) and
is somewhat similar to mikera's [imagez](https://github.com/mikera/imagez).

### Motivation

Collage grew out of my own need when I was writing another Clojure application.
I wanted to have some high-level image processing functionality available -
functions into which I could just pass some `BufferedImages` and get back
transformed `BufferedImages`. At the time, I resorted to doing Java interop,
which is nice, but Clojure is nicer. I also wanted to get more experience with
Clojure in general.

### Project goals

- Ease of use
- Well tested-ness
- Clean, composable internal API
- Learn (more idiomatic) Clojure

Documentation generated by Marginalia is available on
[Github pages](https://karls.github.io/collage/).

Available on [Clojars](https://clojars.org/fivetonine/collage).

## Usage

#### Using the `with-image` macro.

```clj
(:require [fivetonine.collage.util :as util])
(:require [fivetonine.collage.core :refer :all])

(with-image "/path/to/image.jpg"
            (resize :width 1000)
            (rotate 90)
            (util/save :quality 0.85))
```

Loads the image at `/path/to/image.jpg`, resizes it to have width of 1000px
(height is computed automatically), rotates 90 degrees clockwise and saves it
with 85% quality of the original, overwriting the original.

```clj
(:require [fivetonine.collage.util :as util])
(:require [fivetonine.collage.core :refer :all])

(def image (util/load-image "/path/to/image.jpg"))
(with-image image
	        (crop 100 50 200 100)
	        (util/save "/path/to/new-image.jpg" :quality 1.0))
```

Loads an image at `/path/to/image.jpg`. With that image, crops a 200px by 100px
image out of the original, at the point (100,50) in the original image, saves
it with 100% quality at `/path/to/new-image.jpg`.

#### Vanilla functions.

Plain ol' functions

```clj
(:require [fivetonine.collage.util :as util])
(:require [fivetonine.collage.core :refer :all])

(def image (load-image "/path/to/image.png"))
(util/save (flip image :horizontal))
```

or using the [thread-first macro](http://clojuredocs.org/clojure_core/clojure.core/-%3E).

```clj
(:require [fivetonine.collage.util :as util])
(:require [fivetonine.collage.core :refer :all])

(-> (load-image "/path/to/image.png") (flip :horizontal) (util/save :progressive true))
```

## WebP support

Collage does not provide support for WebP out of the box. Collage includes all
the JVM `ImageIO` API plumbing (in resources/webp-imageio.jar), but
the native binaries are not provided. Note that the `webp` format is reported as
a supported format with `ImageIO.getReaderFormatNames()`. But when trying
to load a `.webp` image, an exception is thrown as the native binary for
actually loading the raw data in is missing.

The main reason for not providing binaries is that I don't want to
build and maintain all the versions of all the binaries for all the platforms.

Compiling the binary is fairly straightforward. Some instructions are available
in the luciad's [webp-imageio](https://bitbucket.org/luciad/webp-imageio) repo.
This repo includes code that is used in Collage to provide support for WebP.
I also compiled the native binary that WebP depends on using the instructions
in the same repo.

If you run into problems compiling the binary or getting all the pieces working
together, please open an issue and describe the situation.

WebP support is currently tested only on Mac OSX, but it _should_ work on other
platforms as well.

## Contributing

Contributions, suggestions and friendly criticism are all welcome.

## License

Copyright © 2013 Karl Sutt

Distributed under the Eclipse Public License either version 1.0.


CHANGELOG.md
## 0.3.0

**Added**

- A function to cut a triangle from an image
- A function to cut a circle from an image
- A function to convert an image to grayscale
- A function that adds duotone effect to an image


test/fivetonine/collage/helpers.clj
(ns fivetonine.collage.helpers
  (:import java.awt.image.BufferedImage))

(defn pixel-at
  [image x y]
  (-> image .getData (.getPixel x y (ints nil)) seq))

(defn buf-img
  ([w] (BufferedImage. w w BufferedImage/TYPE_INT_ARGB))
  ([w h] (BufferedImage. w h BufferedImage/TYPE_INT_ARGB))
  ([w h t] (BufferedImage. w h t)))


test/fivetonine/collage/core_test.clj
(ns fivetonine.collage.core-test
  (:require [clojure.test :refer :all]
            [fivetonine.collage.core :refer :all]
            [fivetonine.collage.helpers :refer :all])
  (:import java.awt.image.BufferedImage
           java.awt.Color))

(deftest scale-test
  (let [image (buf-img 100)]
    (testing "scale factor < 1.0"
      (let [scaled (scale image 0.6)]
        (is (= 60 (.getWidth scaled)))
        (is (= 60 (.getHeight scaled)))))

    (testing "scale factor > 1.0"
      (let [scaled (scale image 1.5)]
        (is (= 150 (.getWidth scaled)))
        (is (= 150 (.getHeight scaled)))))

    (testing "scale factor = 1.0"
      (let [scaled (scale image 1.0)]
        (is (= 100 (.getWidth scaled)))
        (is (= 100 (.getHeight scaled)))))))

(deftest crop-test
  (let [image (buf-img 100)]
    (is (= 50 (.getWidth  (crop image 0 0 50 10))))
    (is (= 10 (.getHeight (crop image 0 0 50 10))))
    (is (= 15 (.getWidth  (crop image 50 90 15 5))))
    (is (= 5  (.getHeight (crop image 50 90 15 5))))
    ;; going out of bounds
    (is (thrown? Exception (crop image 90 0 11 10)))
    (is (thrown? Exception (crop image 0 90 10 11)))))

(deftest resize-test
  (let [image (buf-img 800 600)]
    (testing "no width and height provided"
      (is (thrown? IllegalArgumentException (resize image))))

    (testing "with width provided"
      (let [resized (resize image :width 400)]
        (is (= 400 (.getWidth resized)))
        (is (= 300 (.getHeight resized)))))

    (testing "with height provided"
      (let [resized (resize image :height 300)]
        (is (= 400 (.getWidth resized)))
        (is (= 300 (.getHeight resized)))))

    (testing "with width and height both provided"
      (let [resized (resize image :width 50 :height 30)]
        (is (= (.getWidth resized) 50))
        (is (= (.getHeight resized) 30))))))

(deftest resize*-test
  (let [image (buf-img 100)
        resized (resize image :width 50 :height 30)]
    (is (= (.getWidth resized) 50))
    (is (= (.getHeight resized) 30))))

(deftest paste-test
  (let [image (buf-img 10 10)]
    (testing "wrong number of layer args"
      ;; the values of arguments don't matter in the layer vector, as they're
      ;; checked only when it's certain that there is a correct number of them
      (is (thrown? IllegalArgumentException (paste image [:foo :bar])))
      (is (thrown? IllegalArgumentException (paste image [:foo :bar :baz :qux]))))

    (testing "no layers"
      (is (instance? BufferedImage (paste image [])))
      (is (not (= image (paste image [])))))

    (testing "some layers"
      (let [layer1 (buf-img 2)
            graphics1 (.createGraphics layer1)
            layer2 (buf-img 2)
            graphics2 (.createGraphics layer2)]

        (is (instance? BufferedImage (paste image [layer1 0 0])))
        (is (instance? BufferedImage (paste image [layer1 0 0 layer2 5 5])))
        (is (not (= image (paste image [layer1 0 0 layer2 5 5]))))

        (doto graphics1
          (.setPaint (Color/yellow))
          (.fillRect 0 0 2 2)
          (.dispose))
        (doto graphics2
          (.setPaint (Color/green))
          (.fillRect 0 0 2 2)
          (.dispose))

        (let [pasted (paste image [layer1 0 0 layer2 9 9])]
          (is (= (pixel-at pasted 9 0) (pixel-at pasted 0 9)))
          (is (not (= (pixel-at pasted 0 0) (pixel-at pasted 9 9))))
          (is (not (= (pixel-at pasted 0 0) (pixel-at pasted 0 9))))
          (is (not (= (pixel-at pasted 0 9) (pixel-at pasted 9 9)))))))))

(deftest paste*-test
  (let [image (buf-img 10)
        layer (buf-img 5)
        layer-graphics (.createGraphics layer)]
    (doto layer-graphics
      (.setPaint (Color/yellow))
      (.fillRect 0 0 5 5)
      (.dispose))
    (let [pasted (paste* image [layer 0 0])]
      (testing "applying a layer"
        (is (= (pixel-at pasted 0 0) (pixel-at pasted 4 0)))
        (is (not (= (pixel-at pasted 0 0) (pixel-at pasted 5 0))))))))

(deftest rotate-test
  (let [dimensions [640 480]
        image (buf-img (dimensions 0) (dimensions 1))
        new-dims #(let [r (rotate image %1)] [(.getWidth r) (.getHeight r)])]
    (are [theta dim] (= dim (new-dims theta))
         0    dimensions
         90   (reverse dimensions)
         180  dimensions
         270  (reverse dimensions)
         360  dimensions
         450  (reverse dimensions)
         -90  (reverse dimensions)
         -180 dimensions
         -270 (reverse dimensions)
         -360 dimensions
         -450 (reverse dimensions))))


test/fivetonine/collage/util_test.clj
(ns fivetonine.collage.util-test
  (:require [clojure.test :refer :all]
            [clojure.java.io :refer [as-file resource]]
            [clojure.string :as str]
            [fivetonine.collage.util :refer :all]
            [fivetonine.collage.helpers :refer :all])
  (:import java.io.File
           java.awt.image.BufferedImage
           javax.imageio.ImageWriteParam
           java.awt.Color
           javax.imageio.ImageIO
           javax.swing.JFrame
           java.net.URI))

(def test-image-paths
  {"fixtures/images/cloud.png" "fixtures/images/cloud-new.png"
   "fixtures/images/apartment.jpg" "fixtures/images/apartment-new.jpg"})

(defn cleanup-images []
  (doseq [path (vals test-image-paths)]
    (let [file (File. path)]
      (when (.isFile file)
        (.delete file)))))

(deftest load-image-test
  (let [path (first (keys test-image-paths))]
    (are [image] (instance? BufferedImage image)
         (load-image path)
         (load-image (as-file path))
         (load-image (ImageIO/read (as-file path)))
         (load-image (resource (str/replace path #"^fixtures/" ""))))))

(deftest sanitize-path-test
  (is (instance? URI (sanitize-path "file:///path/to/some/image.png")))
  (is (instance? URI (sanitize-path (first (keys test-image-paths)))))
  (is (thrown? Exception (sanitize-path "http://www.foobar.com/image.png"))))

(deftest save-test
  (testing "default options"
    (doseq [[path new-path] (seq test-image-paths)]
      (let [image (load-image path)
            saved-image (save image new-path)]
        (is (.exists (File. saved-image)))))
    (cleanup-images))

  (testing "with progressive mode"
    (testing "turned on"
      (doseq [[path new-path] (seq test-image-paths)]
        (let [image (load-image path)
              saved-image #(save image new-path :progressive true)]
          (is (.exists (File. (saved-image))))))
      (cleanup-images))
    (testing "turned off"
      (doseq [[path new-path] (seq test-image-paths)]
        (let [image (load-image path)
              saved-image #(save image new-path :progressive false)]
          (is (.exists (File. (saved-image)))))))
    (cleanup-images))
  (testing "with explicit quality coefficient"))

(deftest copy-test
  (testing "creating a new object"
    (let [image (BufferedImage. 1 1 BufferedImage/TYPE_INT_ARGB)]
      (is (not (= image (copy image))))))

  (testing "changing old image doesn't change new image"
    (let [image1 (BufferedImage. 1 1 BufferedImage/TYPE_INT_RGB)
          graphics1 (.createGraphics image1)
          image2 (copy image1)
          graphics2 (.createGraphics image2)]
      (doto graphics1
        (.setPaint (Color/yellow))
        (.fillRect 0 0 1 1)
        (.dispose))
      (is (not (= (pixel-at image1 0 0) (pixel-at image2 0 0)))))))


project.clj
(defproject fivetonine/collage "0.3.0"
  :description "Clean, minimal image processing library for Clojure"
  :url "https://github.com/karls/collage"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]]
  :plugins [[lein-marginalia "0.8.0"]]

  :profiles {:test {:resource-paths ["fixtures"]}}

  ;; :global-vars { *warn-on-reflection* true }

  ;; WebP support
  :resource-paths ["resources" "resources/webp-imageio.jar"]
  :java-source-paths ["src/fivetonine/collage/java"]

  ;; In order to actually use the WebP format, the JVM uses native code that
  ;; needs to be compiled by the user. The JVM loads the native code from the
  ;; native library path, which is set here. This may be overridden when
  ;; starting the JVM with -Djava.library.path=/your/custom/path/.
  ;; See the README for instructions on how to compile the native code.
  :jvm-opts [~(str "-Djava.library.path=native/:" (System/getenv "LD_LIBRARY_PATH"))])


src/fivetonine/collage/util.clj
(ns fivetonine.collage.util
  (:require [clojure.java.io :refer [as-file file]])
  (:import java.io.File
           java.net.URI
           java.net.URL
           java.awt.image.BufferedImage
           javax.imageio.ImageIO
           javax.imageio.IIOImage
           javax.imageio.ImageWriter
           javax.imageio.ImageWriteParam
           fivetonine.collage.Frame))

(declare parse-extension)

(defn show
  "Display an image in a `JFrame`.

  Convenience function for viewing an image quickly."
  [^BufferedImage image]
  (Frame/createImageFrame "Quickview" image))

(defn copy
  "Make a deep copy of an image."
  [image]
  (let [width (.getWidth image)
        height (.getHeight image)
        type (.getType image)
        new-image (BufferedImage. width height type)]
    ;; Get data from image and set data in new-image, resulting in a copy
    ;; This also works for BufferedImages that are obtained by calling
    ;; .getSubimage on another BufferedImage.
    (.setData new-image (.getData image))
    new-image))

(defn save
  "Store an image on disk.

  Accepts optional keyword arguments.

  `:quality` - decimal, between 0.0 and 1.0. Defaults to 0.8.

  `:progressive` - boolean, `true` turns progressive saving on, `false`
  turns it off. Defaults to the default value in the ImageIO API -
  `ImageWriteParam/MODE_COPY_FROM_METADATA`. See
  [Java docs](http://docs.oracle.com/javase/7/docs/api/javax/imageio/ImageWriteParam.html).

  Examples:

    (save image \"/path/to/new/image.jpg\" :quality 1.0)
    (save image \"/path/to/new/image/jpg\" :progressive false)

  Returns the path to the saved image when saved successfully."
  [^BufferedImage image path & rest]
  (let [opts (apply hash-map rest)
        outfile (file path)
        ext (parse-extension path)
        ^ImageWriter writer (.next (ImageIO/getImageWritersByFormatName ext))
        ^ImageWriteParam write-param (.getDefaultWriteParam writer)
        iioimage (IIOImage. image nil nil)
        outstream (ImageIO/createImageOutputStream outfile)]
    ; Only compress images that can be compressed. PNGs, for example, cannot be
    ; compressed.
    (when (.canWriteCompressed write-param)
      (doto write-param
        (.setCompressionMode ImageWriteParam/MODE_EXPLICIT)
        (.setCompressionQuality (get opts :quality 0.8))))
    (when (.canWriteProgressive write-param)
      (let [mode-map {true  ImageWriteParam/MODE_DEFAULT
                      false ImageWriteParam/MODE_DISABLED}
            mode-flag (get opts :progressive)]
        (doto write-param
          (.setProgressiveMode (get mode-map
                                    mode-flag
                                    ImageWriteParam/MODE_COPY_FROM_METADATA)))))
    (doto writer
      (.setOutput outstream)
      (.write nil iioimage write-param)
      (.dispose))
    (.close outstream)
    path))

(defprotocol ImageResource
  "Coerce different image resource representations to BufferedImage."
  (as-image [x] "Coerce argument to an image."))

(extend-protocol ImageResource
  String
  (as-image [s] (ImageIO/read (as-file s)))

  File
  (as-image [f] (ImageIO/read f))

  URL
  (as-image [r] (ImageIO/read r))

  BufferedImage
  (as-image [b] b))

(defn ^BufferedImage load-image
  "Loads an image from resource."
  [resource]
  (as-image resource))

;; ## Helpers & experimental

(defn parse-extension
  "Parses the image extension from the path."
  [path]
  (last (clojure.string/split path #"\.")))

(defn sanitize-path
  "Sanitizes a path.
  Returns the sanitized path, or throws if sanitization is not possible."
  [path]
  (when-let [scheme (-> path URI. .getScheme)]
    (if-not (= "file" scheme)
      (throw (Exception. "Path must point to a local file."))
      (URI. path)))
  (URI. (str "file://" path)))


src/fivetonine/collage/core.clj
;; ## Drop in library for (some of) your image processing needs.
;;
;; Collage was developed out of my own need to answer some specific needs in a
;; project I was working on. Even though there are a couple of other libraries
;; out there (mikera's [imagez](https://github.com/mikera/imagez), which builds
;; on [imgscalr](https://github.com/thebuzzmedia/imgscalr), a Java library),
;; but I felt like implementing my own in order to gain more experience with
;; Clojure.
;;
;; The feature-set is somewhat similar to the previously mentioned libraries,
;; adding functionality to paste layers (regular `BufferedImages`) onto an image
;; and controlling the quality of the image when saving it to disk.
;;
;; ### This project aims to
;; * Be an easy to use, drop-in solution
;; * Have a composable internal API
;; * Be well tested
;; * Be reasonably idiomatic
;;
(ns fivetonine.collage.core
  (:require [fivetonine.collage.util :as util])
  (:import java.awt.color.ColorSpace)
  (:import java.awt.image.BufferedImage)
  (:import java.awt.image.ColorConvertOp)
  (:import java.awt.geom.AffineTransform)
  (:import java.awt.AlphaComposite)
  (:import java.awt.Color)
  (:import java.awt.Polygon)
  (:import java.awt.RenderingHints))

(declare resize*)
(declare paste*)
(declare normalise-angle)
(declare pi-rotation?)

(def not-nil? (complement nil?))

;; ## Core functions

(defn rotate
  "Rotates image through angle `theta`, where `theta` is
  an integer multiple of 90.

  If `theta > 0`, the image is rotated clockwise.

  If `theta < 0`, the image is rotated anticlockwise."
  [image theta]
  (when-not (contains? (set (range -360 450 90)) (normalise-angle theta))
    (throw (IllegalArgumentException.
            "theta has to be an integer multiple of 90.")))
  (let [old-width (.getWidth image)
        old-height (.getHeight image)
        new-width (if (pi-rotation? theta) old-width old-height)
        new-height (if (pi-rotation? theta) old-height old-width)
        angle (Math/toRadians theta)
        new-image (BufferedImage. new-width new-height (.getType image))
        graphics (.createGraphics new-image)
        transform (AffineTransform.)]

    ;; Given that the rotation happens around the point (0,0) (the top left hand
    ;; corner of the image), the resulting image needs to be translated back
    ;; into the "viewport".
    (condp = (Math/abs (normalise-angle theta))
      0   (.translate transform 0 0)
      90  (.translate transform new-width 0)
      180 (.translate transform new-width new-height)
      270 (.translate transform 0 new-height)
      360 (.translate transform 0 0))
    (.rotate transform angle)

    (doto graphics
      (.drawImage image transform nil)
      (.dispose))
    new-image))

(defn flip
  "Flips an image.

  If direction is `:horizontal`, flips the image around the y-axis.

  If direction is `:vertical`, flips the image around the x-axis."
  [image direction]
  (let [width (.getWidth image)
        height (.getHeight image)
        new-image (BufferedImage. width height (.getType image))
        graphics (.createGraphics new-image)
        transform (AffineTransform.)]
    (case direction
      :horizontal (doto transform
                    (.translate width 0)
                    (.scale -1 1))
      :vertical (doto transform
                  (.translate 0 height)
                  (.scale 1 -1)))

    (doto graphics
      (.drawImage image transform nil)
      (.dispose))

    new-image))

(defn scale
  "Scales an image by a factor `f`.

  If `0.0 < f < 1.0` the image is scaled down.

  If `f > 1.0` the image is scaled up."
  [image f]
  (resize* image
           (* f (-> image .getWidth int))
           (* f (-> image .getHeight int))))

(defn crop
  "Crops an image.

  `x, y` are the coordinates to top left corner of the area to crop.
  `width, height` are the width and height of the area to crop.

  The returned image does not share its data with the original image."
  [image x y width height]
  (-> image (.getSubimage x y width height) util/copy))

(defn resize
  "Resizes an image.

  If only `width` or `height` is provided, the resulting image will be `width`
  or `height` px wide, respectively. The other dimension will be calculated
  automatically to preserve `width/height` ratio.

  With `width` and `height` both provided, the resulting image will be crudely
  resized to match the provided values.

  If neither `width` nor `height` are provided, `IllegalArgumentException` is
  thrown.

  Examples:

    (resize image :width 100)
    (resize image :height 300)
    (resize image :width 100 :height 300)"
  [image & {:keys [width height] :as opts}]
  (let [supported #{:width :height}
        options (select-keys opts supported)
        width (options :width)
        height (options :height)]
    (when (empty? options)
      (throw (IllegalArgumentException.
              "Width or height (or both) has to be provided.")))
    (cond
     (and width height) (resize* image (int width) (int height))

     (not-nil? width)
     (let [new-height (* (/ (int width) (.getWidth image)) (.getHeight image))]
       (resize* image width (int new-height)))

     (not-nil? height)
     (let [new-width (* (/ (int height) (.getHeight image)) (.getWidth image))]
       (resize* image (int new-width) height)))))

(defn resize*
  "Resize the given image to `width` and `height`.
  Used as an internal function by `resize`.

  Note: the method of resizing may change in the future as there are better,
  iterative, solutions to balancing speed vs. quality. See
  [the perils of Image.getScaledInstance()](https://today.java.net/pub/a/today/2007/04/03/perils-of-image-getscaledinstance.html)."
  [image width height]
  (let [new-image (BufferedImage. width height (.getType image))
        graphics (.createGraphics new-image)]
    (doto graphics
      (.setRenderingHint RenderingHints/KEY_INTERPOLATION
                         RenderingHints/VALUE_INTERPOLATION_BICUBIC)
      (.drawImage image 0 0 width height nil)
      .dispose)
    new-image))

(defn paste
  "Pastes layer(s) onto image at coordinates `x` and `y`.

  `layer-defs` is expected to in the format

    [layer1 x1 y2 layer2 x2 y2 ... ]

  Layers are loaded using `fivetonine.collage.util/load-image`.
  Top left corner of a layer will be at `x, y`.

  Throws `IllegalArgumentException` if the number of elements in the list of
  layers and coordinates is not divisible by 3.

  Returns the resulting image."
  [image & layer-defs]
  (let [args (flatten (seq layer-defs))]
    (when-not (zero? (-> args count (rem 3)))
      (throw (IllegalArgumentException.
              "Expected layer-defs format [image1 x1 y1 image2 x2 y2 ... ].")))
    (let [new-image (util/copy image)
          layers (partition 3 args)]
      ;; "reduce" all the layers onto new-image using paste* -- the layers are
      ;; accumulated onto the image
      (reduce paste* new-image layers)
      new-image)))

(defn paste*
  "Paste layer on top of base at position `x, y` and return the resulting image.
  Used as an internal function by `paste`."
  [base [layer x y]]
  (let [graphics (.createGraphics base)]
    (doto graphics
      (.setRenderingHint RenderingHints/KEY_RENDERING
                         RenderingHints/VALUE_RENDER_QUALITY)
      (.setRenderingHint RenderingHints/KEY_COLOR_RENDERING
                         RenderingHints/VALUE_COLOR_RENDER_QUALITY)
      (.drawImage (util/load-image layer) x y nil)
      (.dispose))
    base))

(defn circle
  "Crop a circle from an image, leaving a transparent background. This function
  does not compose well with other functions that cut from the image, as its
  implementation doesn't really cut - it paints parts of the image onto a
  circle. If you need to compose circles with other shapes, create the circle
  first.

  Why must this function be so limited? Using a clip path yields a much too
  jagged circle. Instead we paint a beautifully anti-aliased circle, and then
  composit the image on top. This yields a pleasantly anti-aliased circle image
  with the drawback that if you apply the transformation to a transparent image,
  you will see the parts of the white circle the image doesn't cover."
  [^BufferedImage image]
  (let [width (.getWidth image)
        circle (BufferedImage. width width BufferedImage/TYPE_INT_ARGB)]
    (doto (.createGraphics circle)
      (.setComposite AlphaComposite/Clear)
      (.fillRect 0 0 width width)
      ;; Fake soft-clipping by first drawing the desired clip shape in fully
      ;; opaque white antialiasing enabled
      (.setComposite AlphaComposite/Src)
      (.setRenderingHint RenderingHints/KEY_ANTIALIASING RenderingHints/VALUE_ANTIALIAS_ON)
      (.setColor (Color/WHITE))
      (.fill (java.awt.geom.Ellipse2D$Float. 0 0 width width))
      ;; Then composit the image on top
      (.setComposite AlphaComposite/SrcAtop)
      (.drawImage image 0 0 nil)
      .dispose)
    circle))

(defn- triangle-polygon [pos ^long w]
  (case pos
    :lower-left (Polygon. (int-array [0 w 0]) (int-array [w w 0]) 3)
    :lower-right (Polygon. (int-array [w w 0]) (int-array [w 0 w]) 3)
    :upper-right (Polygon. (int-array [w 0 w]) (int-array [0 0 w]) 3)
    :upper-left (Polygon. (int-array [0 0 w]) (int-array [0 w 0]) 3)))

(defn triangle
  "Crop a triangle from an image, leaving a transparent background. Can optionally
  be called with an option map that specifies where the triangle is positioned:

  ```clj
  (triangle image {:position :upper-left})
  (triangle image {:position :lower-left})
  (triangle image {:position :lower-right})
  (triangle image {:position :upper-right})
  ```"
  [^BufferedImage image & [{:keys [position]}]]
  (let [width (.getWidth image)
        triangle (BufferedImage. width width BufferedImage/TYPE_INT_ARGB)]
    (doto (.createGraphics triangle)
      (.setComposite AlphaComposite/Clear)
      (.fillRect 0 0 width width)
      (.setComposite AlphaComposite/Src)
      (.setClip (triangle-polygon (or position :upper-left) width))
      (.drawImage image 0 0 width width nil)
      .dispose)
    triangle))

(defn grayscale
  "Convert an image to grayscale."
  [^BufferedImage image]
  (let [output (BufferedImage. (.getWidth image) (.getHeight image) (.getType image))]
    (-> (ColorConvertOp. (.. image getColorModel getColorSpace) (ColorSpace/getInstance ColorSpace/CS_GRAY) nil)
        (.filter image output))
    output))

(defmacro with-image
  "A helper for applying multiple operations to an image.

  `image-resource` can be a `String`, a `File` or a `BufferedImage`.

  Example:

    (with-image \"/path/to/image.jpg\"
                 (scale 0.8)
                 (rotate 90)
                 (crop 0 0 100 100))

  Expands to (properly namespaced):

    (let [image__2336__auto__ (load-image \"/path/to/image.jpg\")]
      (clojure.core/-> image__2336__auto__
                       (scale 0.8)
                       (rotate 90)
                       (crop 0 0 100 100)))

  Returns the image which is the result of applying all operations to the input
  image."
  [image-resource & operations]
  `(let [image# (util/load-image ~image-resource)]
     (-> image# ~@operations)))

;; ## Helpers

(defn- pi-rotation?
  "Does the rotation through angle `theta` correspond to a rotation that is
  a multiple of 180 degrees, meaning that the image preserves the original
  width and height."
  [theta]
  (-> theta (/ 90) (rem 2) (= 0)))

(defn normalise-angle
  "Restrict the rotation angle to the range [-360..360]."
  [theta]
  (rem theta 360))
